File: 2015/13/sol.js
const fs = require('fs');
const itertools = require('itertools')


function ImportFile(fileName)
{
    let lines =  (fs.readFileSync(fileName, 'utf8')).split('\n'); 

    let data = { p : new Set(), edges : new Object() };

    for (const line of lines)
    {
        let a = line.split(' ')[0];
        let b = line.split(' ')[10].split('.')[0];
        data.p.add(a);
        data.p.add(b);
        if (!data.edges[a]) data.edges[a] = new Object();

        if (line.split(' ')[2] == 'gain') 
        {
            data.edges[a][b] = parseInt(line.split(' ')[3]);
        }
        else 
        {
            data.edges[a][b] = -parseInt(line.split(' ')[3]);
        }
    }


    data.edges['me'] = new Object();
    for (const p of data.p)
    {
        data.edges[p]['me'] = 0;
        data.edges['me'][p] = 0;        
    }

    data.p.add('me');
    return data;
}

data = ImportFile('text.txt');




function Seat(P1 = true)
{
    let max = 0;
    let perms = [];
    if (P1 ) perms = itertools.permutations([...data.p].filter(x => x != 'me'));
    else  perms = itertools.permutations([...data.p]);
    for (let perm of perms)
    {
        perm.push(perm[0]);
        sum = 0;
        for (let i = 0; i < perm.length -1; i++)
        {
            let a = perm[i];
            let b = perm[(i+1)];
            sum += data.edges[a][b] + data.edges[b][a];
        }
        if (sum > max) max = sum;
    }
    return max;
}   


console.log(Seat());
console.log(Seat(false));

File: 2015/17/sol.py
import itertools
v = 150


def ImportFile(r):
    f = open(r, 'r')
    data = dict()
    i = 0
    for _ in f.read().split('\n'):
        data[i] = int(_)
        i += 1

    return data


def P1(data):
    keys = list(data.keys())
    nums = dict()
    perms = list(itertools.combinations(keys, len(keys)))
    for i in range(1, len(keys) - 1):
        perms += list(itertools.combinations(keys, i))
    combos = 0
    for c in perms:
        sum = 0
        for e in c:
            sum += data[e]
        if sum == v:
            combos += 1
            if (not len(c) in nums.keys()):
                nums[len(c)] = 1
            else:
                nums[len(c)] += 1

    return (combos, nums[min(list(nums.keys()))])


data = ImportFile('text.txt')
result = P1(data)
print(result[0])
print(result[1])
